beginner_source/mnist_train_nas.py

```diff
@@ -0,0 +1,170 @@
+        hidden_layers = [self.hidden_size_1, self.hidden_size_2]
+        num_params = 0
+        for hidden_size in hidden_layers:
+            if hidden_size > 0:
+                layers.append(nn.Linear(width, hidden_size))
+                layers.append(nn.ReLU())
+                layers.append(nn.Dropout(self.dropout))
+                num_params += width * hidden_size
+                width = hidden_size
+        layers.append(nn.Linear(width, self.num_classes))
+        num_params += width * self.num_classes
+
+        # Save the model and parameter counts
+        self.num_params = num_params
+        self.model = nn.Sequential(*layers)  # No need to use Relu for the last layer
+
+    def forward(self, x):
+        x = self.model(x)
+        return F.log_softmax(x, dim=1)
+
+    def training_step(self, batch, batch_idx):
+        x, y = batch
+        logits = self(x)
+        loss = F.nll_loss(logits, y)
+        return loss
+
+    def validation_step(self, batch, batch_idx):
+        x, y = batch
+        logits = self(x)
+        loss = F.nll_loss(logits, y)
+        preds = torch.argmax(logits, dim=1)
+        acc = accuracy(preds, y)
+        self.log("val_acc", acc, prog_bar=False)
+        return loss
+
+    def configure_optimizers(self):
+        optimizer = torch.optim.Adam(self.parameters(), lr=self.learning_rate)
+        return optimizer
+
+    def prepare_data(self):
+        MNIST(self.data_dir, train=True, download=True)
+        MNIST(self.data_dir, train=False, download=True)
+
+    def setup(self, stage=None):
+        self.mnist_train = MNIST(self.data_dir, train=True, transform=self.transform)
+        self.mnist_val = MNIST(self.data_dir, train=False, transform=self.transform)
+
+    def train_dataloader(self):
+        return DataLoader(self.mnist_train, batch_size=self.batch_size)
+
+    def val_dataloader(self):
+        return DataLoader(self.mnist_val, batch_size=self.batch_size)
+
+
+mnist_model = MnistModel()
+
+
+# Initialize a trainer (don't log anything since things get so slow...)
+trainer = Trainer(
+    logger=False,
+    gpus=AVAIL_GPUS,
+    max_epochs=args.epochs,
+    enable_progress_bar=False,
+    deterministic=True,  # Do we want a bit of noise?
+    default_root_dir=args.log_path,
+)
+
+logger = pl_loggers.TensorBoardLogger(args.log_path)
+
+eprint(args.log_path)
+
+# Train the model and log time ⚡
+start = time.time()
+trainer.fit(model=mnist_model)
+end = time.time()
+train_time = end - start
+logger.log_metrics({"train_time": end - start})
+
+# Compute the validation accuracy once and log the score
+with io.capture_output() as captured:
+    val_accuracy = trainer.validate()[0]["val_acc"]
+logger.log_metrics({"val_acc": val_accuracy})
+
+# Log the number of model parameters
+num_params = trainer.model.num_params
+logger.log_metrics({"num_params": num_params})
+
+logger.save()
+
+# Print outputs
+eprint(f"train time: {train_time}, val acc: {val_accuracy}, num_params: {num_params}")
```

Comment:
Can this be refactored into main routine and guarded with `if __name__ == "main":`?
```suggestion
logger = pl_loggers.TensorBoardLogger(args.log_path)

eprint(args.log_path)

# Train the model and log time ⚡
start = time.time()
trainer.fit(model=mnist_model)
end = time.time()
train_time = end - start
logger.log_metrics({"train_time": end - start})

# Compute the validation accuracy once and log the score
with io.capture_output() as captured:
    val_accuracy = trainer.validate()[0]["val_acc"]
logger.log_metrics({"val_acc": val_accuracy})

# Log the number of model parameters
num_params = trainer.model.num_params
logger.log_metrics({"num_params": num_params})

logger.save()

# Print outputs
eprint(f"train time: {train_time}, val acc: {val_accuracy}, num_params: {num_params}")
```